src/Microsoft.OData.Core/ODataBatchReader.cs

```diff
@@ -203,49 +203,24 @@ void IODataBatchOperationListener.BatchOperationContentStreamDisposed()
+        /// Continues reading from the batch message payload.
         /// </summary>
-        /// <returns>The next state of the batch reader.</returns>
-        private ODataBatchReaderState GetEndBoundaryState()
-        {
-            switch (this.batchReaderState)
-            {
-                case ODataBatchReaderState.Initial:
-                    // If we find an end boundary when in state 'Initial' it means that we
-                    // have an empty batch. The next state will be 'Completed'.
-                    return ODataBatchReaderState.Completed;
-
-                case ODataBatchReaderState.Operation:
-                    // If we find an end boundary in state 'Operation' we have finished
-                    // processing an operation and found the end boundary of either the
-                    // current changeset or the batch.
-                    return this.batchStream.ChangeSetBoundary == null
-                        ? ODataBatchReaderState.Completed
-                        : ODataBatchReaderState.ChangesetEnd;
-
-                case ODataBatchReaderState.ChangesetStart:
-                    // If we find an end boundary when in state 'ChangeSetStart' it means that
-                    // we have an empty changeset. The next state will be 'ChangeSetEnd'
-                    return ODataBatchReaderState.ChangesetEnd;
-
-                case ODataBatchReaderState.ChangesetEnd:
-                    // If we are at the end of a changeset and find an end boundary
-                    // we reached the end of the batch
-                    return ODataBatchReaderState.Completed;
-
-                case ODataBatchReaderState.Completed:
-                    // We should never get here when in Completed state.
-                    throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataBatchReader_GetEndBoundary_Completed));
-
-                case ODataBatchReaderState.Exception:
-                    // We should never get here when in Exception state.
-                    throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataBatchReader_GetEndBoundary_Exception));
-
-                default:
-                    // Invalid enum value
-                    throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataBatchReader_GetEndBoundary_UnknownValue));
-            }
-        }
+        /// <returns>true if more items were read; otherwise false.</returns>
+        internal abstract bool ReadImplementation();
+
+        /// <summary>
+        /// Returns the cached <see cref="ODataBatchOperationRequestMessage"/> for reading the content of an operation 
+        /// in a batch request.
+        /// </summary>
+        /// <returns>The message that can be used to read the content of the batch request operation from.</returns>
+        internal abstract ODataBatchOperationRequestMessage CreateOperationRequestMessageImplementation();

```

Comment:
I was thinking about using protected too, which would make this API accessible from outside of the assembly (e.g. possible for a public client to derive ODataMyBatchReader). It appears to me that internal is providing better restriction. Can you share the related guidance in OData for using protected/internal?

---
In reply to: [108274517](https://github.com/OData/odata.net/pull/779#discussion_r108274517) [](ancestors = 108274517)